packages/workspace/src/browser/workspace-service.ts

```diff
@@ -102,31 +206,119 @@ export class WorkspaceService implements FrontendApplicationContribution {
+                preserveWindow: this.preferences['workspace.preserveWindow'] || !(rootToOpen),
                 ...options
             };
             await this.server.setRoot(rootUri);
             if (preserveWindow) {
-                this._root = valid;
+                await this.setupRoots();
             }
-            this.openWindow(uri, { preserveWindow });
+            await this.openWindow({ preserveWindow });
+            this.updateTitle();
             return;
         }
         throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootUri}.`);
     }
 
     /**
-     * Clears current workspace root and reloads window.
+     * Adds a root folder to the workspace
+     * @param uri URI of the root folder being added
+     */
+    async addRoot(uri: URI): Promise<void> {
+        const workspaceRootChange = await this.workspaceRootChange;
+        const workspaceFolder = workspaceRootChange.root;
+        if (!this.opened || !workspaceFolder) {
+            throw new Error('Folder cannot be added as there is no active folder in the current workspace.');
+        }
+
+        const rootToAdd = uri.toString();
+        const valid = await this.toValidRoot(rootToAdd);
+        if (valid) {
+            if (!this._workspaceRootChange.roots.find(r => r.uri === valid.uri)) {
+                await this.server.addRoot(rootToAdd, workspaceFolder.uri);
+                this._workspaceRootChange.roots.push(valid);
+                const configPath = await this.server.getWorkspaceConfigFilePath(this._root!.uri);
+                if (configPath) {
+                    this._workspaceConfigFile = await this.fileSystem.getFileStat(configPath);
+                } else {
+                    this._workspaceConfigFile = undefined;
+                }
+                this._workspaceRootChange.workspaceConfigFile = this._workspaceConfigFile;
+                this.startWatch(valid);
+                this.onWorkspaceRootChangeEmitter.fire(this._workspaceRootChange);
+            }
+            return;
+        }
+        throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootToAdd}.`);
+    }
+
+    /**
+     * Removes folder(s) from workspace.
+     */
+    async removeFolders(uris: URI[]): Promise<void> {
+        if (!this.opened) {
+            throw new Error('Folder cannot be removed as there is no active folder in the current workspace.');
+        }
+        const validUris: string[] = [];
+        for (const uri of uris.map(u => u.toString())) {
+            const valid = await this.toValidRoot(uri);
+            if (valid) {
+                validUris.push(uri);
+            }
+        }
+
+        if (validUris.length > 0 && this._root) {
+            const messageContainer = document.createElement('div');
+            messageContainer.textContent = 'Do you really want to remove the following roots and their subfolders from the workspace?';
```

Comment:
Would some users be scared that the content of the removed roots might be deleted? Since there is already an appropriate title on the dialog (Remove Folder From Workspace), maybe we could shorten the message and mention nothing is physically erased? e.g.:
"Remove the following folders from workspace? (note: nothing will be erased from disk)"